java/core/src/java/org/apache/orc/impl/ParserUtils.java

```diff
@@ -244,80 +250,172 @@ public static TypeDescription parseType(ParserUtils.StringPosition source) {
 
   private static final Pattern INTEGER_PATTERN = Pattern.compile("^[0-9]+$");
 
+  public static ColumnVector[] findBranchVectors(TypeDescription schema,
+                                                 ParserUtils.StringPosition source,
+                                                 VectorizedRowBatch batch) {
+    List<String> names = ParserUtils.splitName(source);
+    return vectorWalker.get().find(schema, names, true, batch).vectorBranch;
+  }
+
   public static TypeDescription findSubtype(TypeDescription schema,
                                             ParserUtils.StringPosition source) {
     return findSubtype(schema, source, true);
   }
 
-  public static TypeDescription findSubtype(TypeDescription schema,
-                                            ParserUtils.StringPosition source,
-                                            boolean isSchemaEvolutionCaseAware) {
-    List<String> names = ParserUtils.splitName(source);
-    if (names.size() == 1 && INTEGER_PATTERN.matcher(names.get(0)).matches()) {
-      return schema.findSubtype(Integer.parseInt(names.get(0)));
+  private static final ThreadLocal<TypeWalker> typeWalker = ThreadLocal.withInitial(TypeWalker::new);
+  private static class TypeWalker {
+    private List<String> names;
+    private boolean caseAware;
+
+    // Walk specific
+    TypeDescription current = null;
+
+    private TypeWalker() {
     }
-    TypeDescription current = SchemaEvolution.checkAcidSchema(schema)
-        ? SchemaEvolution.getBaseRow(schema) : schema;
-    while (names.size() > 0) {
-      String first = names.remove(0);
-      switch (current.getCategory()) {
-        case STRUCT: {
-          int posn = -1;
-          if (isSchemaEvolutionCaseAware) {
-            posn = current.getFieldNames().indexOf(first);
-          } else {
-            // Case-insensitive search like ORC 1.5
-            for (int i = 0; i < current.getFieldNames().size(); i++) {
-              if (current.getFieldNames().get(i).equalsIgnoreCase(first)) {
-                posn = i;
-                break;
+
+    protected void handleLevel(int posn) {
+      this.current = current.getChildren().get(posn);
+    }
+
+    private void findInternal() {
```

Comment:
I think I'd prefer to leave the findInternal as a static method that takes a visitor. I've posted a branch with such a change on my github. https://github.com/omalley/orc/tree/orc-755

Reply:
Will check and revert, thanks for the proposal

Reply:
I have incorporated the suggestion with some tweaks.